Client/Frontend/UserContent/UserScripts/RequestBlocking.js

```diff
@@ -0,0 +1,100 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+((args) => {
+  "use strict";
+  //[DEBUG]console.log('[RB] Loaded request blocking')
+  const securityToken = args.securityToken
+  const sendMessage = (resourceURL, resourceType) => {
+    //[DEBUG]console.log(`[RB] Checking if request should be blocked for '${resourceURL}'`)
+    return webkit.messageHandlers[args.handlerName].postMessage({
+      securityToken: securityToken,
+      resourceURL: resourceURL.href,
+      sourceURL: window.location.href,
+      resourceType: resourceType
+    })
+  }
+  
+  const makeURLFromString = (urlString) => {
+    try {
+      return new URL(urlString, window.location.href)
+    }
+    catch(err) {
+      //[DEBUG]console.log(`[RB] URL Error: '${err}'`)
+      return
+    }
+  }
+  
+  const { fetch: originalFetch } = window
+  window.fetch = function() {
+    const [resource, config] = arguments
+    
+    // Extract the url string
+    let urlString
+    if (typeof resource === 'string' || resource instanceof String) {
+      // Simple send where resource is a string
+      urlString = resource
+    } else if (resource !== undefined && (typeof resource.url === 'string' || resource.url instanceof String)) {
+      // Assume its a Request
+      urlString = resource.url
+    } else {
+      //[DEBUG]console.log('[RB] Cannot extract url')
+      return originalFetch.apply(this, arguments)
+    }
+
+    // Resolve the url in case it's a relative path
+    const url = makeURLFromString(urlString)
+    if (url === undefined) { return }
+    //[DEBUG]console.log(`[RB] URL: '${url}'`)
+    
+    return sendMessage(url, 'script').then(blocked => {
+      if (blocked) {
+        //[DEBUG]console.log(`[RB] Request BLOCKED for '${url}'`)
+        return new Promise(function(_, reject) {
+          reject(Error(`Content blocked for '${url}'`))
+        });
+      } else {
+        return originalFetch.apply(this, arguments)
+      }
+    })
+  }
+  
+  const originalOpen = XMLHttpRequest.prototype.open
+  XMLHttpRequest.prototype.open = function() {
+    const sendArguments = arguments
+    originalOpen.apply(this, arguments)
+
+    // Extract the url string
+    let urlString = arguments[1]
+    if (!(typeof urlString === 'string' || urlString instanceof String)) {
```

Comment:
why the typechecking here?  What else are you expecting than a string?  And why check for string twice?